perfmetrics/scripts/hns_rename_folders_metrics/generate_folders_and_files.py

```diff
@@ -95,7 +95,85 @@ def list_directory(path) -> list:
     return contents_url
   except subprocess.CalledProcessError as e:
     logmessage(e.output.decode('utf-8'))
-    subprocess.call('bash', shell=True)
+
+
+def compare_folder_structure(folder, folder_url) -> bool:
+  """Checks if the number of files in config file under folder matches the
+  number of files in the GCS bucket.
+
+  Args:
+    folder: Json Object representing the folder.
+    folder_url: Corresponding folder url in the GCS bucket.
+
+  Returns:
+    true if the number of files in the folder in GCS bucket matches the num_files
+    parameter of the folder JSON object.
+    false otherwise
+  """
+  try:
+    files_in_folder = list_directory(folder_url)
+    if len(files_in_folder) != folder["num_files"]:
+      return False
+  except:
+    # If the list directory fails wth url did not match object, folder
+    # specified in config file does not exist in bucket.
+    return False
+
+  return True
+
+
+def check_if_dir_structure_exists(directory_structure) -> bool:
+  """Checks if the directory structure mentioned in the config file already
+  exists in the GCS bucket.
+
+  Args:
+    directory_structure: Json Object representing the directory structure.
+
+  Returns:
+    true if the existing structure in GCS bucket exactly matches with teh config
+     file
+    false otherwise
+  """
+  bucket_name = directory_structure["name"]
+  bucket_url = 'gs://{}'.format(bucket_name)
+
+  # Check for top level folders.
+  folders = list_directory(bucket_url)
+  nested_folder_count = "nested_folders" in directory_structure
+  if "folders" in directory_structure:
+    if len(folders) != directory_structure["folders"][
+      "num_folders"] + nested_folder_count:
+      return False
+
+    # For each non-nested folder , check the count of files.
+    for folder in directory_structure["folders"]["folder_structure"]:
+      folder_url = '{}/{}'.format(bucket_url, folder["name"])
+      match = compare_folder_structure(folder, folder_url)
+      if not match:
+        return False
+
+  # Check the number of second level folders in nested folders.
+  if nested_folder_count:
+    nested_folder = directory_structure["nested_folders"]["folder_name"]
+    try:
+      second_level_folders = list_directory(
```

Comment:
The logic of comparing top level and nested folders look same. Can we extract the common logic into a function and re-use?

Reply:
The logic slightly differs due to the occasional existence of nested_folder folder in the top level directory.
For the top level folder, we have some folders which just contain files and (optional) nested folder which itself contains folders with files.
Inside  the nested folder check, we do not have to deal with this case as all the second level folders are simple folders with just files.

Reply:
Misunderstood.Added new function with common functionality.